---
title: "Untitled"
author: "Ruth Kristianingsih"
date: "10/12/2019"
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_dir = here::here("reports", stringr::str_remove(getwd(), here::here("scripts/r-scripts/")))) })
output:
  md_document:
    variant: markdown_github
---

# Getting secreted data

## Introduction

Ideally, the heatmap plot made in jupyter lab book can show which positions of the sequences activated the most by convolutional layers, however after grouping the data based on the pathogen organisms (it still has not shows similar pattern for each organism). Hypothesis: This may due to the effector data which is not secreted by the organism. 

## Aim

In this report, several steps will be conducted in order to get the protein data secreted from each pathogen organisms (to improve the perfomance of the model in recognising the pattern). 

## Plan

In order to do that, several plans will be done:


  1. Using `rentrez` library, get the genomic sequence ID by inputting the GeneID known from data we have (Phi-base data). 
  
  2. If the the genomic sequence ID exist, then:
  
  - After getting the genomic sequence ID, then get all of protein sequences encoded from there. 
  
  However, in many cases, the protein does not have genomic sequence ID:
  
  - If it is the case, then I may need to pick the protein randomly from the same pathogen organism and species.
  
  3. After getting the sequence, we need to use SignalP to identify the signal peptide, this way, we know whether a protein is secreted or not. The threshold of probability to be accepted to have signal peptide = 90%. 
  

## Execution

### Load of the data

```{r include = FALSE}
# Load libraries
library(tidyverse)
library(rentrez)
library(ape)
```

```{r}
# Loading the data

# Load the effector data obtained previously
effector_data <- data.table::fread("../../../data/getting-data-new/binary-class-data/effector_data.csv")

# Load phi base data
all_phi_base <- data.table::fread("../../../data/phi-base-current-data/phi-base_without_column_desc.csv")
```

Get all of the ProteinID from the effector data. 

```{r}
list_protein_id <- effector_data %>%
  pull(ProteinID)
```

```{r}
# Getting the information about the Gene names and Gene IDs, and the protein source
phi_base_prot <- all_phi_base %>%
  dplyr::filter(ProteinID %in% list_protein_id) %>%
  group_by(ProteinID, ProteinIDsource) %>%
  slice(1) %>%
  dplyr::select(ProteinIDsource, ProteinID, GeneIDsource, GeneID, Gene, PathogenID, Pathogenspecies, PathogenstrainID, Pathogenstrain)


# Getting the dictinct value of each data
phi_base_prot %>%
  dplyr::distinct(Gene) %>%
  nrow()
```

From the result, it seems that the they have unique name. NOw, using the unique name we have above, we want to get the genome name. 

The example of the genome I can get from gene is `KGO57615` ==> input in Gene, and then I can get the info of the Genome, and using that Genome ID, retrive it in Genome NCBI. Get the whole protein sequences. In this case, using the gene `LysM1` will not result in a specific one. 

### Making sure we connect to internet

```{r}
# Function to check whether we connect to internet
connected.to.internet <- function() {
  if (curl::curl_fetch_memory("www.google.com")$status_code == 200) {
    return(TRUE)
  } else {
    stop(
      "It seems that you are not connected to the internet. Could you please check?",
      call. = FALSE
    )
  }
}

connected.to.internet()
```

Below is a function to get the information about the genomic sequences by accessing the html file. 

```{r}
# get_proteins_in_genome <- function(gene){
#   # URL of the queries:
#   #   https://www.ncbi.nlm.nih.gov/genome/?term=<QUERY>
#   # URL to get the Genome ID if the gene ID is known
#   #   https://www.ncbi.nlm.nih.gov/gene/?term=<GENE_ID>
#   # URL to get all protein data in all genome
#   #   https://www.ncbi.nlm.nih.gov/nuccore/<GNOME_ID>
#
#   # gene_name <- paste0(gene)
#   # Get the gene information
#   # Base URL and queries
#   base_url <- "https://www.ncbi.nlm.nih.gov"
#   search_gene_query <- "/gene?term="
#
#   # Download HTML file
#   search_gene_url <- paste0(base_url, search_gene_query, gene)
#   search_gene_html_file <- textreadr::read_html(search_gene_url)
#
#   # Get the genomic sequence ID
#   genomic_sequence_id <- search_gene_html_file %>%
#   stringr::str_detect("Genomic Sequence") %>%
#   which() %>%
#   `+`(1) %>%
#   search_gene_html_file[.]
#
#   # Get the all of sequence from the whole genomic sequence
#   search_whole_genome_query <- "/nuccore/"
#   search_whole_genome_url <- paste0(base_url, search_whole_genome_query, genomic_sequence_id)
#   search_whole_genome_html_file <- textreadr::read_html(search_whole_genome_url)
#
#   print(search_whole_genome_url)
#   # return(search_whole_genome_html_file)
# }
```

### Checking if a protein has Gene and Gene ID or not

```{r}
# Function if protein ID has protein ID
check_if_genomic_seq_exist <- function(gene_term) {
  if (nchar(gene_term) > 0) {
    # Get the Gene ID from NCBI
    ncbi_gene_id <- entrez_search(db = "gene", term = gene_term)$ids

    # Return boolean whether the sequence exists or not
    seq_exists <- (length(ncbi_gene_id) != 0)

    return(seq_exists)
  } else {
    return(NA)
  }
}
```


```{r}
df_check_genomic_seq <- phi_base_prot %>%
  ungroup() %>%
  rowwise() %>%
  dplyr::mutate("ncbi_genomic_seq_exist" = check_if_genomic_seq_exist(GeneID))
```

```{r}
df_check_genomic_seq %>% 
  group_by(ncbi_genomic_seq_exist) %>% 
  summarise(count = n()) %>% 
  knitr::kable()
```




From the table above, we can see that more than a half of the data does not have the genomic sequence. Therefore, we will first get the protein sequences for the data that does have genomic sequence. 

We can now get the ID of the gene terms with the genomic sequence exists in NCBI.

```{r}
# Get the list of geneID with the genomic sequence exist in NCBI
geneID_with_genomic_seq <- df_check_genomic_seq %>% 
  dplyr::filter(ncbi_genomic_seq_exist == TRUE) %>% 
  pull(GeneID)

geneID_with_genomic_seq
```

```{r}
# Get the list of geneID with the genomic sequence does not exist in NCBI
geneID_without_genomic_seq <- df_check_genomic_seq %>% 
  dplyr::filter(ncbi_genomic_seq_exist == FALSE) %>% 
  pull(GeneID)

geneID_without_genomic_seq
```


```{r}
gene_with_genomic_seq_NA <- df_check_genomic_seq %>% 
  dplyr::filter(is.na(ncbi_genomic_seq_exist)) %>% 
  pull(Gene)

gene_with_genomic_seq_NA
```

### Retriving genomic sequence ID for the proteins that have Gene ID from phi_base data

```{r}
# Define a function to get the protein ids if the Gene ID is known
get_protein_NCBI_id <- function(gene_id){
  
  # Retrive the genomic seq NCBI ID id the Gene ID is known 
  get_genomic_id <- entrez_link(dbfrom='protein', id = gene_id , db='all')$links$protein_nuccore
  
  # Get all of protein IDs using 
  protein_ids <- entrez_link(dbfrom = "nuccore", id = get_genomic_id, db = "all")$links$nuccore_protein
  
  return(protein_ids)
}
```

```{r}
get_protein_NCBI_id("KGO57615")
```


### Get the protein sequences from genomic sequences

```{r}
# Function to predict the secreted protein, with the probability that the signal peptide exist = 90%
get_signalp_pred <- function(fasta_filename, verbose = FALSE) {
  signalp_path <- "/Users/kristian/Documents/Workspace/Software/signalp/bin"
  verbose_string <- tolower(deparse(substitute(verbose)))
  fasta_path <- fasta_filename

  command_string <- paste0(
    "export PATH=", signalp_path, ":$PATH ;",
    " signalp -fasta ", fasta_path, " -org euk -format short ",
    "-verbose=", verbose_string, " -stdout"
  )

  data_lines <- command_string %>%
    system(intern = TRUE)

  data <- data_lines %>%
    .[-c(1, 2)] %>%
    as.list() %>%
    data.frame() %>%
    t() %>%
    as_tibble() %>%
    tidyr::separate(col = V1, into = c("ID", "Prediction", "SP(Sec/SPI)", "OTHER", "CS Position"), sep = "\t")

  return(data)
}
```



```{r}
get_all_prot_seqs <- function(gene_id) {

  # Get the protein IDs
  ncbi_all_protein_id <- get_protein_NCBI_id(gene_id)

  # Get all of the protein sequences
  prot_fasta <- entrez_fetch(db = "protein", id = ncbi_all_protein_id, rettype = "fasta")

  # Save in temp file
  temp <- tempfile()
  write(prot_fasta, temp)
  parsed_fasta <- ape::read.FASTA(temp, type="AA")

  # Get the SignalP prediction from the fasta file we get
  pred <- get_signalp_pred(temp)

  # Get the information about fasta file if it is secreted
  pred_signalP_exist <- pred %>%
    # dplyr::filter(Prediction == "SP(Sec/SPI)") %>%
    dplyr::mutate(Gene = gene_id)

  return(pred_signalP_exist)
}
```

```{r}
get_all_prot_seqs("EXK24251")
```

```{r}
list_ids_without_NA <- c(geneID_with_genomic_seq, geneID_without_genomic_seq)
```


```r
# Initialize 
out = NULL
# Create for loop to get the protein contained in the same genomic sequence
for (i in 1:3){
  
    get_prot <- get_all_prot_seqs(list_ids_without_NA)
    
    out = rbind(out, get_prot)
}
```



## Get information the negative sequence for protein data that do not have neither genomic sequence  in NCBI nor the geneID on the list

```{r}
eff_info_without_NCBI_GeneID <- phi_base_prot %>% 
  dplyr::filter(Gene %in% gene_with_genomic_seq_NA) 

eff_info_without_NCBI_GeneID
```

```{r}
eff_info_without_NCBI_GeneID <- eff_info_without_NCBI_GeneID %>% 
  ungroup() %>% 
  select("PathogenID", "PathogenstrainID")

eff_info_without_NCBI_GeneID <- eff_info_without_NCBI_GeneID %>% 
  dplyr::mutate(pathogen_strain_id = ifelse(!is.na(PathogenstrainID), PathogenstrainID, PathogenID))

eff_info_without_NCBI_GeneID_ids <- eff_info_without_NCBI_GeneID %>% 
  pull(pathogen_strain_id)
```

```{r}
eff_info_without_NCBI_GeneID_ids
```

```{r eval=FALSE, include=FALSE}
get_txid_data <- function(txid) {
  # Construct query
  txid_query <- paste0("txid", txid, "[Organism]")
  txid_base_query <- "NOT virulence[All Fields] NOT effector[All Fields] NOT elicitor[All Fields] NOT partial[All Fields] NOT multispecies[All Fields] NOT \"Unknown\"[Organism] NOT hypothetical[All Fields] NOT uncharacterized[All Fields] NOT unnamed[All Fields] NOT putative[All Fields]"
  txid_full_query <- paste(txid_query, txid_base_query)

  # Perform query
  search <- rentrez::entrez_search(db = "protein", term = txid_full_query, use_history = TRUE)
  records <- rentrez::entrez_fetch(db = "protein", web_history = search$web_history, retmax = search$count, rettype = "fasta")

  # Split and clean individual records
  records_clean <- records %>%
    stringr::str_split("\n>") %>%
    unlist() %>%
    stringr::str_replace(">", "") %>%
    unlist()

  # Get ID and sequence
  # To working within the list structure, and keep each record separated from one another,
  # we use lapply(x, fun) (apply function fun to each element of list)
  clean_txid_data <- data.frame(
    txid = txid,
    # Get ID
    id = records_clean %>%
      # Get first element of FASTA
      stringr::str_split(" ") %>%
      lapply(`[`, 1) %>%
      # Remove possible NAs
      lapply(function(x) {
        stringr::str_replace_na(x, "")
      }) %>%
      unlist(),
    # Get sequence
    sequence = records_clean %>%
      # Get sequence from FASTA
      stringr::str_split("\n") %>%
      lapply(`[`, 2:length(.)) %>%
      # Remove possible NAs
      lapply(function(x) {
        stringr::str_replace_na(x, "")
      }) %>%
      # Collapse sequence into a single string
      lapply(function(x) {
        stringr::str_c(x, collapse = "")
      }) %>%
      unlist()
  ) %>%
    # Clean ID
    dplyr::rowwise() %>%
    dplyr::mutate(
      # Force id to be a character
      id = as.character(id),
      # Clean strange something|REALID|something ids
      id = ifelse(
        stringr::str_detect(id, "\\|"),
        stringr::str_split(id, "\\|") %>% unlist() %>% .[2],
        id
      ),
    ) %>%
    # Get rid of unusable data
    dplyr::filter(
      id != "",
      !is.na(id)
    ) %>%
    ungroup()

  return(clean_txid_data)
}
```

```{r eval=FALSE, include=FALSE}
get_txid_data_from_txid_list <- function(txid_list) {
  # Empty data frame
  tx_full <- data.frame(
    rowid = numeric(),
    txid = character(),
    id = character(),
    sequence = character()
  )

  # Add new data from txid_list
  for (rowid in 1:length(txid_list)) {
    tx_full <- tx_full %>%
      rbind(
        cbind(
          rowid = rowid,
          get_txid_data(txid_list[rowid])
        )
      )
  }

  return(tx_full)
}
```


## Check the group of organism of effector data

```{r}
# Checking the organism name

bacteria <- c("Acinetobacter baumannii", 
"Aeromonas hydrophila", 
"Aeromonas salmonicida", 
"Brucella abortus", 
"Burkholderia glumae", 
"Burkholderia pseudomallei", 
"Campylobacter jejuni", 
"Citrobacter rodentium", 
"Clavibacter michiganensis", 
"Coxiella burnetii", 
"Cystobacter fuscus", 
"Edwardsiella ictaluri", 
"Erwinia amylovora", 
"Escherichia coli", 
"Francisella tularensis", 
"Helicobacter pylori", 
"Legionella pneumophila",  
"Listeria monocytogenes", 
"Mycobacterium tuberculosis", 
"Pantoea stewartii", 
"Pseudomonas aeruginosa", 
"Pseudomonas cichorii", 
"Pseudomonas savastanoi", 
"Pseudomonas syringae", 
"Ralstonia solanacearum", 
"Salmonella enterica", 
"Shigella flexneri", 
"Staphylococcus aureus", 
"Vibrio parahaemolyticus", 
"Xanthomonas axonopodis", 
"Xanthomonas campestris", 
"Xanthomonas oryzae", 
"Xylella fastidiosa", 
"Yersinia enterocolitica", 
"Yersinia pestis", 
"Yersinia pseudotuberculosis")

fungi <- c("Beauveria bassiana",
"Blumeria graminis",
"Botrytis cinerea",
"Cercospora apii",
"Cercospora beticola",
"Colletotrichum orbiculare",
"Dothistroma septosporum",
"Fusarium oxysporum",
"Leptosphaeria maculans",
"Magnaporthe oryzae",
"Melampsora lini",
"Parastagonospora nodorum",
"Passalora fulva",
"Penicillium expansum",
"Pseudocercospora fuligena",
"Puccinia striiformis",
"Rhynchosporium commune",
"Verticillium dahliae",
"Ustilago maydis",
"Zymoseptoria tritici")

oomycetes <- c("Hyaloperonospora arabidopsidis", 
"Phytophthora cactorum", 
"Phytophthora capsici", 
"Phytophthora infestans", 
"Phytophthora parasitica", 
"Phytophthora sojae", 
"Pythium aphanidermatum", 
"Plasmopara halstedii", 
"Phytophthora megakarya")

others <- c("Globodera rostochiensis",
"Heterodera glycines",
"Macrosiphum euphorbiae",
"Toxoplasma gondii")
```

#### Fungi

```{r}
phi_base_prot_with_organism <- phi_base_prot %>% 
  rowwise() %>% 
  dplyr::mutate(class = case_when(
  (Pathogenspecies %in% bacteria) ~ "bacteria",
  (Pathogenspecies %in% fungi) ~ "fungi",
  (Pathogenspecies %in% oomycetes) ~ "oomycete",
  TRUE ~ "others"))
```

```{r}
phi_base_prot_with_organism %>% 
  dplyr::filter(class == "fungi") %>% 
  dplyr::select(c("ProteinID", "GeneID", "Gene", "PathogenID", "Pathogenspecies", "class")) %>% 
  group_by(Pathogenspecies) %>% 
  summarise(count = n())
```
Some of the pathogen species that are not available in the biomart:

Cercospora apii			
Cercospora beticola
Melampsora lini
Passalora fulva
Penicillium expansum
Pseudocercospora fuligena
Rhynchosporium commune



Note that Phaeosphaeria nodorum (synonym and correct taxonomic name: Parastagonospora nodorum) -- Wikipedia. 

#### Bacteria 

In different .Rmd

```{r}
phi_base_prot_with_organism_bacteria <- phi_base_prot_with_organism %>% 
  dplyr::filter(class == "bacteria") 
  # dplyr::select(c("ProteinID", "GeneID", "Gene", "PathogenID", "Pathogenspecies", "class")) %>% 
  # group_by(Pathogenspecies) %>% 
  # summarise(count = n())
```

```{r}
prot_bacteria_ID <- phi_base_prot_with_organism_bacteria %>% 
  pull(ProteinID)

write.csv(prot_bacteria_ID, "prot_bacteria_ID.csv")
```

```{r}
prot_bacteria_with_ensembl_reference <- data.table::fread("prot_bacteria_with_ensembl.tab") %>% 
  dplyr::select(c(Entry, "EnsemblBacteria transcript")) %>% 
  `colnames<-`(c("ProteinID", "EnsemblBacteria")) %>% 
  rowwise() %>% 
  mutate(EnsemblBacteria = stringr::str_remove_all(EnsemblBacteria, ";"))
``` 

```{r}
prot_gene_pathogen_bacteria <- phi_base_prot_with_organism_bacteria %>% 
  dplyr::select(c(ProteinID, GeneID, Pathogenspecies))
```



Getting organism name from the GeneID in EnsemblBacteria

```{r}
get_organism_name <- function(geneID){
  # URL of the queries:
  #   http://ensemblgenomes.org/id/<QUERY>

  # Get the gene information
  # Base URL and queries
  base_url <- "http://ensemblgenomes.org/id/"

  # Download HTML file
  search_name_url <- paste0(base_url, geneID)
  search_name_html_file <- textreadr::read_html(search_name_url)

  # # Get the genomic sequence ID
  name_organism <- search_name_html_file %>%
  stringr::str_detect("Species") %>%
  which() %>%
  `+`(1) %>%
  search_name_html_file[.]
  # 
  # # Get the all of sequence from the whole genomic sequence
  # search_whole_genome_query <- "/nuccore/"
  # search_whole_genome_url <- paste0(base_url, search_whole_genome_query, genomic_sequence_id)
  # search_whole_genome_html_file <- textreadr::read_html(search_whole_genome_url)
  
  print(geneID)
  print(name_organism)
  # return(name_organism)
  # return(search_whole_genome_html_file)
}
```

```{r}
get_organism_name("CAN00074")
```

```{r}
# Function to get the name of each organism in a table 

prot_bacteria_with_ensemblbacteria_organism <-prot_gene_pathogen_bacteria %>% 
  rowwise() %>% 
  dplyr::mutate(organism_name = ifelse(!is.null(GeneID), get_organism_name(GeneID), Pathogenspecies))
```



#### Oomycete

```{r}
phi_base_prot_with_organism %>% 
  dplyr::filter(class == "oomycete") %>% 
  dplyr::select(c("ProteinID", "GeneID", "Gene", "PathogenID", "Pathogenspecies", "class")) %>% 
  group_by(Pathogenspecies) %>% 
  summarise(count = n())
```

Not in Ensembl:
Phytophthora cactorum		
Phytophthora capsici












