---
title: "Untitled"
author: "Ruth Kristianingsih"
date: "08/01/2020"
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_dir = here::here("reports", stringr::str_remove(getwd(), here::here("scripts/r-scripts/")))) })
output:
  md_document:
    variant: markdown_github
---

# Getting secreted Organism name for bacteria


## Load the libraries 

```{r}
library(tidyverse)
```


```{r}
# Loading the data

# Load the effector data obtained previously
effector_data <- data.table::fread("../../../data/getting-data-new/binary-class-data/effector_data.csv")

# Load phi base data
all_phi_base <- data.table::fread("../../../data/phi-base-current-data/phi-base_without_column_desc.csv")
```


```{r}
effector_data_class <- readRDS("../../../data/getting-data-new/multi-class-data/class_df_effectors.rds")
```

```{r}
sapply(effector_data_class, class)
```

```{r}
sapply(effector_data, class)
```

```{r}
effector_data <- effector_data %>% 
  dplyr::mutate(PathogenID = as.character(PathogenID))
```


```{r}
effector_data_complete <- effector_data %>% 
  left_join(effector_data_class %>% ungroup(), by ="PathogenID")
```




Get all of the ProteinID from the effector data. 

```{r}
list_protein_id <- effector_data_complete %>%
  pull(ProteinID)
```

```{r}
# Getting the information about the Gene names and Gene IDs, and the protein source
phi_base_prot <- all_phi_base %>%
  dplyr::filter(ProteinID %in% list_protein_id) %>%
  group_by(ProteinID, ProteinIDsource) %>%
  slice(1) %>%
  dplyr::select(ProteinIDsource, ProteinID, GeneIDsource, GeneID, Gene, PathogenID, Pathogenspecies, PathogenstrainID, Pathogenstrain)

# Getting the dictinct value of each data
phi_base_prot %>%
  dplyr::distinct(Gene) %>%
  nrow()
```
```{r}
# Checking the organism name

# bacteria <- c("Acinetobacter baumannii", 
# "Aeromonas hydrophila", 
# "Aeromonas salmonicida", 
# "Brucella abortus", 
# "Burkholderia glumae", 
# "Burkholderia pseudomallei", 
# "Campylobacter jejuni", 
# "Citrobacter rodentium", 
# "Clavibacter michiganensis", 
# "Coxiella burnetii", 
# "Cystobacter fuscus", 
# "Edwardsiella ictaluri", 
# "Erwinia amylovora", 
# "Escherichia coli", 
# "Francisella tularensis", 
# "Helicobacter pylori", 
# "Legionella pneumophila",  
# "Listeria monocytogenes", 
# "Mycobacterium tuberculosis", 
# "Pantoea stewartii", 
# "Pseudomonas aeruginosa", 
# "Pseudomonas cichorii", 
# "Pseudomonas savastanoi", 
# "Pseudomonas syringae", 
# "Ralstonia solanacearum", 
# "Salmonella enterica", 
# "Shigella flexneri", 
# "Staphylococcus aureus", 
# "Vibrio parahaemolyticus", 
# "Xanthomonas axonopodis", 
# "Xanthomonas campestris", 
# "Xanthomonas oryzae", 
# "Xylella fastidiosa", 
# "Yersinia enterocolitica", 
# "Yersinia pestis", 
# "Yersinia pseudotuberculosis")
# 
# fungi <- c("Beauveria bassiana",
# "Blumeria graminis",
# "Botrytis cinerea",
# "Cercospora apii",
# "Cercospora beticola",
# "Colletotrichum orbiculare",
# "Dothistroma septosporum",
# "Fusarium oxysporum",
# "Leptosphaeria maculans",
# "Magnaporthe oryzae",
# "Melampsora lini",
# "Parastagonospora nodorum",
# "Passalora fulva",
# "Penicillium expansum",
# "Pseudocercospora fuligena",
# "Puccinia striiformis",
# "Rhynchosporium commune",
# "Verticillium dahliae",
# "Ustilago maydis",
# "Zymoseptoria tritici")
# 
# oomycetes <- c("Hyaloperonospora arabidopsidis", 
# "Phytophthora cactorum", 
# "Phytophthora capsici", 
# "Phytophthora infestans", 
# "Phytophthora parasitica", 
# "Phytophthora sojae", 
# "Pythium aphanidermatum", 
# "Plasmopara halstedii", 
# "Phytophthora megakarya")
# 
# others <- c("Globodera rostochiensis",
# "Heterodera glycines",
# "Macrosiphum euphorbiae",
# "Toxoplasma gondii")
```

```{r}
# phi_base_prot_with_organism <- phi_base_prot %>% 
#   rowwise() %>% 
#   dplyr::mutate(class = case_when(
#   (Pathogenspecies %in% bacteria) ~ "bacteria",
#   (Pathogenspecies %in% fungi) ~ "fungi",
#   (Pathogenspecies %in% oomycetes) ~ "oomycete",
#   TRUE ~ "others"))
```

From the phi base protein, we can get the information about the pathogen strain for each sequence. 

```{r}
sapply(phi_base_prot, class)
```

```{r}
# Change the class of pathogenID 

phi_base_prot <- phi_base_prot %>% 
  dplyr::mutate(PathogenID = as.character(PathogenID))
```


```{r}
effector_data_info <- phi_base_prot %>% 
  ungroup() %>% 
  left_join(effector_data_complete, by = "ProteinID") %>% 
  dplyr::select(c("ProteinID", "GeneID", "Gene", "PathogenID.x", "Pathogenspecies.x", "Pathogenstrain", "class")) %>% 
  `colnames<-` (c("ProteinID", "GeneID", "Gene", "PathogenID", "Pathogenspecies", "Pathogenstrain", "class"))


effector_data_info %>% 
  saveRDS(., "effector_data_info.RDS")
```


```{r}
effector_data_info_bacteria <- effector_data_info %>% 
  dplyr::filter(class == "bacteria") 
```


```{r}
effector_data_info_bacteria %>% 
  group_by(Pathogenspecies) %>% 
  summarise(count = n()) %>% 
  knitr::kable()
```



#### Getting organism name from EnsemblBacteria
Getting organism name from the GeneID in EnsemblBacteria


```{r}
get_organism_name <- function(geneID) {
  # URL of the queries:
  #   http://ensemblgenomes.org/id/<QUERY>

  # Make sure if there's any query
  if (nchar(geneID) == 0) {
    return(NA)
  }
  # Get the gene information
  # Base URL and queries
  base_url <- "http://ensemblgenomes.org/id/"

  # Download HTML file
  search_name_url <- paste0(base_url, geneID)
  search_name_html_file <- textreadr::read_html(search_name_url)

  # message(paste("Querying", geneID))
  # Get the genomic sequence ID
  name_organism <- search_name_html_file %>%
    stringr::str_detect("Species") %>%
    which() %>%
    `+`(1) %>%
    search_name_html_file[.]
  
  if (length(name_organism) == 0) {
    name_organism <- NA
  }

  # message(geneID, " ", name_organism)
  return(name_organism)
}
```


```{r}
get_organism_name("CAN00074")
```

```{r}
# Function to get the name of each organism in a table 

effector_bacteria_with_ensemblbacteria_organism <- effector_data_info_bacteria %>% 
  rowwise() %>% 
  dplyr::mutate(Pathogenspecies_ensemble = ifelse(!is.null(GeneID), get_organism_name(GeneID), Pathogenspecies))
```

```{r}
# If using gene, we cant find the oragnism_name, then altenatively we just need to combine the pathogenspecies and pathogenstrain name

# bacteria_with_organism_name <- prot_bacteria_with_ensemblbacteria_organism %>% 
effector_bacteria_with_ensemblbacteria_organism <- effector_bacteria_with_ensemblbacteria_organism %>% 
  dplyr::mutate(
    Pathogenspecies_ensemble = case_when(
      is.na(Pathogenspecies_ensemble) & !is.null(Pathogenstrain) ~ paste(Pathogenspecies, Pathogenstrain, sep = " "),
      TRUE ~ Pathogenspecies_ensemble
    )) 
```

```{r}
effector_bacteria_with_ensemblbacteria_organism %>% 
  group_by(Pathogenspecies_ensemble) %>% 
  summarise(count = n()) %>% 
  write.csv(., "effector_bacteria_with_ensemblbacteria_organism.csv")
  # knitr::kable()
```

```{r}
effector_complete <- readRDS("../../../data/secreted_data/data_processed_after_signalp/effector_complete_three_pathogens.RDS")

effector_complete %>% 
  dplyr::group_by(class) %>% 
  summarise(count = n())
```

```{r}
pathogen_species_final %>% 
  knitr::kable()
```


```{r}
pathogen_species_final %>% 
  mutate(Patheogenspceies_ensemble_final = stringr::str_to_lower(Patheogenspceies_ensemble_final)) %>%
  group_by(Patheogenspceies_ensemble_final) %>% 
  summarise(count_total = sum(count)) %>% 
  nrow() %>% 
  knitr::kable()
```






After checking one by on on the FTP list of bacteria on ensebml bacteria, there are some organisms that are not available on  the database, which are:

```{r}
unavailable_data <- c("Erwinia amylovora",
"Erwinia amylovora Ea1189",
"Pseudomonas syringae pv. glycinea race 3",
"Pseudomonas syringae pv. glycinea race 4",
"Pseudomonas syringae pv. maculicola M2",
"Pseudomonas syringae pv. morsprunorum, strain R1-5244",
"Pseudomonas syringae pv. phaseolicola 1302A (race 4)",
"Pseudomonas syringae pv. phaseolicola 1448A",
"Pseudomonas syringae pv. phaseolicola 1449A race5 & 7",
"Pseudomonas syringae pv. phaseolicola 1449B (race7)",
"Pseudomonas syringae pv. phaseolicola NPS3121",
"Pseudomonas syringae pv. phaseolicola race 3",
"Pseudomonas syringae pv. pisi 151",
"Pseudomonas syringae pv. pisi 870A",
"Pseudomonas syringae pv. tabaci",
"Pseudomonas syringae pv. tomato JL1065",
"Pseudomonas syringae pv. tomato PT23",
"Ralstonia solanacearum RS1000",
"Ralstonia solanacearum ZJ3721",
"Xanthomonas oryzae pv. oryzae BXO1"
) %>% 
  stringr::str_to_lower()
```

```{r}
test <- bacteria_with_organism_name_sum %>% 
  ungroup() %>% 
  mutate(organism_name = stringr::str_to_lower(organism_name)) %>% 
  dplyr::filter(
  organism_name %>%
    stringr::str_detect(
      paste0(
        "^",
        unavailable_data %>%
          stringr::str_c(collapse = "|"),
        "$"
      ),
      negate = TRUE
    )
)
```

```{r}
library(stringr)
```


```{r}
bacteria_with_organism_name_sum %>% 
  dplyr::filter(!organism_name %in% unavailable_data) %>% 
  dplyr::mutate(organism_name = stringr::str_to_lower(organism_name)) %>% 
  group_by(organism_name) %>% 
  summarise(total = sum(count)) %>% 
  knitr::kable()
```

