---
title: "Untitled"
author: "Ruth Kristianingsih"
date: "17/05/2019"
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_dir = here::here("reports", stringr::str_remove(getwd(), here::here("scripts/r-scripts/")))) })
output:
  md_document:
    variant: markdown_github
---

# Getting the additional data for each datasets

```{r setup, echo=FALSE, message=FALSE}
library(tidyverse)
```


According to the blast results we obtained previously, there are some data that need to be removed since they are identical with other protein sequences. 

## Load the data for each datasets that has been removed

```{r}
df_train <- data.table::fread("../../../data/getting-data-old/df_training_new_after_removed.csv", drop = "V1")
df_val <- data.table::fread("../../../data/getting-data-old/df_validation_new_after_removed.csv", drop = "V1")
df_test <- data.table::fread("../../../data/getting-data-old/df_testing_new_after_removed.csv", drop = "V1")
```

## Load fasta data

```{r}
parse_fasta_data_ncbi <- function(file_path) {
  # Read FASTA file
  fasta_data <- seqinr::read.fasta(file_path)
  # Number of entries
  num_data <- fasta_data %>% length()


  # Create empty data frame
  parsed_data <- data.frame(
    protein_id = rep(NA, num_data),
    protein_fun = rep(NA, num_data),
    pathogen = rep(NA, num_data),
    sequence = rep(NA, num_data)
  )

  for (i in 1:num_data) {
    # Read 'Annot' attribute and parse the string between 'OS=' and 'OX='
    pathogen <- fasta_data[[i]] %>%
      attr("Annot") %>%
      sub(".*\\[ *(.*?) *\\].*", "\\1", .)

    protein_id <- fasta_data[[i]] %>%
      attr("name")

    protein_fun <- fasta_data[[i]] %>%
      attr("Annot") %>%
      stringr::str_remove(protein_id) %>%
      sub(".*> *(.*?) *\\[.*", "\\1", .)

    # Concatenate the vector of the sequence into a single string
    sequence <- fasta_data[[i]] %>%
      as.character() %>%
      toupper() %>%
      paste(collapse = "")

    # Input values into data frame
    parsed_data[i,] <- cbind(protein_id, protein_fun, pathogen, sequence)
  }

  return(parsed_data)
}
```

```{r}
# path 
add_effector_path <- "../../../data/getting-data-old/BLAST-data/0002-additional-data-fasta/batch_entrez_effector.fasta"
add_noneffector_path <- "../../../data/getting-data-old/BLAST-data/0002-additional-data-fasta/batch_entrez_noneffector.fasta"

add_effector_parsed <- parse_fasta_data_ncbi(add_effector_path)
add_noneffector_parsed <- parse_fasta_data_ncbi(add_noneffector_path)

# 
add_noneffector <- add_noneffector_parsed %>% 
  select(sequence) %>% 
  mutate(label = as.factor(0))

add_effector <- add_effector_parsed %>% 
  select(sequence) %>% 
  mutate(label = as.factor(1))

all_removed_rows_freq <- data.table::fread("../../../data/getting-data-old/all_removed_rows_freq.csv")

all_removed_rows_freq %>%
  knitr::kable()
```

```{r}
# Add the data of non-effector and effector to the training data

df_train <- df_train %>%
  rbind(., add_noneffector[1:76, ]) %>% 
  rbind(., add_effector[1:21, ])

df_val <- df_val %>% 
  rbind(., add_noneffector[77:112, ]) %>% 
  rbind(., add_effector[22:27, ])

df_test <- df_test %>% 
  rbind(., add_noneffector[113:140, ]) %>% 
  rbind(., add_effector[28:38, ])

# write.csv(df_train, "../../../data/getting-data-old/BLAST-data/0003-new-data-sets/training-data.csv", col.names = TRUE)
# write.csv(df_val, "../../../data/getting-data-old/BLAST-data/0003-new-data-sets/validation-data.csv", col.names = TRUE)
# write.csv(df_test, "../../../data/getting-data-old/BLAST-data/0003-new-data-sets/testing-fata.csv", col.names = TRUE)
```
